File: src/test/java/agents/starter.java
package agents;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

/**
 * Control your bots in order to destroy the enemy team !
 **/
public
class starter {
    static String orders = "";
    static Set<Bot> enemyBots = new HashSet<>();
    static Set<Bot> allyBots = new HashSet<>();
    static final String ALLY = "ALLY";
    static final String ENEMY = "ENEMY";
    static final String ATTACK = "ATTACK";
    static final String MOVE = "MOVE";
    static final String IDLE = "IDLE";
    static final String FLEE = "FLEE";

    private static class BotInfo {
        int rangeFromBot;
        int distBotRank;
        int shieldComp;
        int healthComp;
        int totalComp;

        BotInfo(int rangeFromBot, int distBotRank, int shieldComp, int healthComp, int totalComp) {
            this.rangeFromBot = rangeFromBot;
            this.distBotRank = distBotRank;
            this.shieldComp = shieldComp;
            this.healthComp = healthComp;
            this.totalComp = totalComp;
        }
    }

    private static class Bot {
        int id;
        String team;
        int shield;
        int health;
        String action;
        private final IntStream targetIds;
        Set<Bot> targets;
        int distEn;
        int distMe;
        int distMeRank;
        int distEnRank;
        int shieldRank;
        int totalRank;
        int borderDist;
        int borderDistRank;
        Set<Bot> attackingMe = new HashSet<>();
        Set<Bot> fleeingMe = new HashSet<>();
        Set<Bot> approachingMe = new HashSet<>();
        Map<Integer, BotInfo> infoFromMyPerspective = new HashMap<>();

        Bot(int id, String team, int shield, int health, String action, String targets, int distEn, int distMe, int distMeRank, int distEnRank, int shieldRank, int totalRank, int borderDist, int borderDistRank) {
            this.id = id;
            this.team = team;
            this.shield = shield;
            this.health = health;
            this.action = action;
            this.targetIds = Arrays.stream(targets.split(",")).mapToInt(Integer::parseInt);
            this.distEn = distEn;
            this.distMe = distMe;
            this.distMeRank = distMeRank;
            this.distEnRank = distEnRank;
            this.shieldRank = shieldRank;
            this.totalRank = totalRank;
            this.borderDist = borderDist;
            this.borderDistRank = borderDistRank;
        }

        public void getTargets(Map<Integer, Bot> bots) {
            // If one of the target is dead, we add a deadBot to the targets
            this.targets = this.targetIds.mapToObj(id -> bots.getOrDefault(id, deadBot(id))).collect(Collectors.toSet());
            this.targets.forEach(bot -> bot.addTargetingMe(this));
        }

        public void attack(int targetId) {
            orders += String.format("%d %s %d;", id, ATTACK, targetId);
        }

        private void movement(String action, int... targetIds) {
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < targetIds.length; i++) {
                sb.append(targetIds[i]);
                if (i < targetIds.length - 1) {
                    sb.append(",");
                }
            }
            orders += String.format("%d %s %s;", id, action, sb);
        }

        private void movement(String action, Bot... targetIds) {
            movement(action, Arrays.stream(targetIds).mapToInt(bot -> bot.id).toArray());
        }

        public void move(int... targetIds) {
            movement(MOVE, targetIds);
        }

        public void move(Bot... targetBots) {
            movement(MOVE, targetBots);
        }

        public void flee(int... targetIds) {
            movement(FLEE, targetIds);
        }

        public void flee(Bot... targetBots) {
            movement(FLEE, targetBots);
        }

        public void idle() {
            orders += String.format("%d %s;", id, IDLE);
        }

        public void addTargetingMe(Bot targetingMe) {
            switch (targetingMe.action) {
                case ATTACK:
                    this.attackingMe.add(targetingMe);
                    break;
                case FLEE:
                    this.fleeingMe.add(targetingMe);
                    break;
                case MOVE:
                    this.approachingMe.add(targetingMe);
                    break;
            }
        }

        public void addInfoFromMyPerspective(int botId, BotInfo info) {
            this.infoFromMyPerspective.put(botId, info);
        }

        public BotInfo viewedBy(Bot bot) {
            return bot.infoFromMyPerspective.get(this.id);
        }

        public Bot getClosestEnemy() {
            assert this.team.equals(ALLY) : "You can only call closestEnemy method on an ally bot";
            return enemyBots.stream().min(Comparator.comparingInt(bot -> bot.viewedBy(this).distBotRank)).get();
        }

        public boolean isDead() {
            return this.health == -1;
        }
    }

    static Bot deadBot(int id) {
        return new Bot(id, "", 0, -1, "", "", 0, 0, 0, 0, 0, 0, 0, 0);
    }

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int botPerPlayer = in.nextInt(); // the amount of bot you control
        int arenaSize = in.nextInt(); // the arena size in meters (just here in case the creator change the arena size during the contest)

        // game loop
        while (true) {
            orders = "";
            int allyBotAlive = in.nextInt(); // the amount of your bot which are still alive
            int botCount = in.nextInt(); // the amount of bots in the arena
            Map<Integer, Bot> bots = new HashMap<>();
            allyBots = new HashSet<>();
            enemyBots = new HashSet<>();
            for (int i = 0; i < botCount; i++) {
                int botId = in.nextInt(); // the unique bot id, stay the same for the whole game
                String botType = in.next(); // the bot type in a string. It can be ALLY | ENEMY
                int health = in.nextInt(); // the approximate bot health. Can be 0 | 25 | 50 | 75 | 100, 25 meaning that your life is >= 25% and < 50% of your max life
                int shield = in.nextInt(); // the approximate bot shield. Can be 0 | 1 | 25 | 50 | 75 | 100, 1 meaning that your shield is >= 1% and < 25% of your max shield and 0 that you have no more shield left
                String action = in.next(); // action executed by the bot last turn
                String targets = in.next(); // list of the targets id targeted by the bot last turn ("id1;id2;id3...") if the bot is a bot, else -1 (the target for IDLE is the bot itself)
                int enemyRange = in.nextInt(); // /!\ALWAYS 0 FOR THE FIRST LEAGUE/!\ then, the RANGE from to the closest enemy (if the bot is an enemy it returns the range to its closest ally)
                int borderRange = in.nextInt(); // /!\ALWAYS 0 FOR THE FIRST LEAGUE/!\ approximate distance between the bot and the closest border
                int borderDistRank = in.nextInt(); // /!\ALWAYS 0 FOR THE FIRST 2 LEAGUES/!\ bot are sorted in ascending order based on their distance to the closest border
                int distEnRank = in.nextInt(); // /!\ALWAYS 0 FOR THE FIRST 2 LEAGUES/!\ bots are sorted by ascending order based on their distance to the closest enemy
                int healthRank = in.nextInt(); // /!\ALWAYS 0 FOR THE FIRST 2 LEAGUES/!\ bots are sorted in ascendant order based on their amount of health, this is the rank of the current bot in the sorted list
                int shieldRank = in.nextInt(); // /!\ALWAYS 0 FOR THE FIRST 2 LEAGUES/!\ bots are sorted in ascendant order based on their amount of shield
                int totalRank = in.nextInt(); // /!\ALWAYS 0 FOR THE FIRST 2 LEAGUES/!\ bots are sorted in ascendant order based on their amount of health + shield
                Bot bot = new Bot(botId, botType, shield, health, action, targets, enemyRange, distEnRank,
                        healthRank, distEnRank, shieldRank, totalRank, borderRange, borderDistRank);
                bots.put(botId, bot);
                if (botType.equals(ALLY)) {
                    allyBots.add(bot);
                } else {
                    enemyBots.add(bot);
                }
            }
            bots.values().forEach(bot -> bot.getTargets(bots));
            for (int i = 0; i < allyBotAlive; i++) {
                in.nextLine(); // skip to the next line
                int onAirId = Integer.parseInt(in.nextLine().split(" ")[0]); // the id of the ON AIR bot
                for (int j = 0; j < botCount - 1; j++) {
                    int botId = in.nextInt(); // the unique bot id
                    in.next(); // the bot type in a string. It can be ALLY | ENEMY
                    int rangeFromOnAirBot = in.nextInt(); // approximate distance between the target and the ON AIR bot. Can be 0 to 3 for short, medium, long and out of range
                    int distMeRank = in.nextInt(); // bots are sorted by ascending order based on their distance to the ON AIR bot
                    int shieldComp = in.nextInt(); // /!\ALWAYS 0 FOR THE FIRST LEAGUE/!\ -1 if the bot has more shield than the ON AIR bot, 0 if it's equal, 1 if the ON AIR bot has more shield
                    int healthComp = in.nextInt(); // /!\ALWAYS 0 FOR THE FIRST LEAGUE/!\ same as shieldComp but for the health
                    int totComp = in.nextInt(); // /!\ALWAYS 0 FOR THE FIRST LEAGUE/!\ same as shieldComp but based on the sum of health+shield
                    bots.get(onAirId).addInfoFromMyPerspective(botId, new BotInfo(rangeFromOnAirBot, distMeRank, shieldComp, healthComp, totComp));
                }
            }
            // All bots move to the closest enemy
            for (Bot allyBot : allyBots) {
                allyBot.move(allyBot.getClosestEnemy());
            }
            System.err.println(allyBots);

            // Write an answer using System.out.println()
            // To debug: System.err.println("Debug messages...");

            System.out.println(orders);
        }
    }
}

File: starterAIs/starter.py
import sys

enemyBots = []
ally_bots = []
bots = {}
ALLY = "ALLY"
ENEMY = "ENEMY"
ATTACK = "ATTACK"
MOVE = "MOVE"
IDLE = "IDLE"
FLEE = "FLEE"
SHORT_RANGE = 0
MEDIUM_RANGE = 1
LONG_RANGE = 2
OUT_OF_RANGE = 3

# Control your bots in order to destroy the enemy team !
bot_per_player = int(input())  # the amount of bot you control
# the arena size in meters (just here in case the creator change the arena size during the contest)
arena_size = int(input())
orders = ""


class BotInfo:
    def __init__(self, range_from_bot, dist_bot_rank, shield_comp, health_comp, total_comp):
        self.range_from_bot = range_from_bot
        self.dist_bot_rank = dist_bot_rank
        self.shield_comp = shield_comp
        self.health_comp = health_comp
        self.total_comp = total_comp

    def __str__(self):
        return "rangeFromBot: {}, distBotRank: {}, shieldComp: {}, healthComp: {}, totalComp: {}".format(
            self.range_from_bot, self.dist_bot_rank, self.shield_comp, self.health_comp, self.total_comp)


def dead_bot(bot_id):
    return Bot(bot_id, "", -1, -1, "", "", -1, -1, -1, -1, -1, -1, -1)


class Bot:
    # Decorator to handle Bot and id for function argument, ignore the warnings of the IDE
    def __bot_or_id(f):
        # noinspection PyCallingNonCallable
        def wrapper(self, *args):
            return f(self, *[arg.bot_id if isinstance(arg, Bot) else arg for arg in args])

        return wrapper

    def __init__(self, bot_id, team, shield, health, action, targets, enemy_range, dist_en_rank, health_rank,
                 shield_rank, total_rank, border_range, border_dist_rank):
        self.bot_id = bot_id
        self.team = team
        self.shield = shield
        self.health = health
        self.action = action
        self.targets = map(int, targets.split(","))
        self.enemy_range = enemy_range
        self.dist_en_rank = dist_en_rank
        self.health_rank = health_rank
        self.shield_rank = shield_rank
        self.total_rank = total_rank
        self.border_range = border_range
        self.border_dist_rank = border_dist_rank
        self.attacking_me = set()
        self.fleeing_me = set()
        self.approaching_me = set()
        self.info_from_my_perspective = {}

    def __str__(self):
        # we can't display all attributes of a bot, so we display only the ones we need
        return f"{self.team} bot {self.bot_id} with {self.health} health and {self.shield} shield"

    def get_targets(self):
        # If one of the target is dead, we add a deadBot to the targets
        self.targets = list(map(lambda bot_id: bots.get(bot_id, dead_bot(bot_id)), self.targets))
        for target in self.targets:
            target.add_targeting_me(self)

    def add_targeting_me(self, targeting_me):
        if targeting_me.action == ATTACK:
            self.attacking_me.add(targeting_me)
        elif targeting_me.action == FLEE:
            self.fleeing_me.add(targeting_me)
        elif targeting_me.action == MOVE:
            self.approaching_me.add(targeting_me)

    @__bot_or_id  # now attack is a function that takes a bot or an id as argument
    def attack(self, target):
        global orders
        orders += f"{self.bot_id} {ATTACK} {target};"

    def __movement(self, action, *targets):
        global orders
        f = lambda bot: str(bot.bot_id) if isinstance(bot, Bot) else str(bot)
        orders += f"{self.bot_id} {action} {','.join(map(f, targets))};"

    def move(self, targets):
        self.__movement(MOVE, targets)

    def flee(self, targets):
        self.__movement(FLEE, targets)

    def idle(self):
        global orders
        orders += f"{self.bot_id} {IDLE};"

    def add_info_from_my_perspective(self, bot_id, info):
        self.info_from_my_perspective[bot_id] = info

    def viewed_by(self, bot):
        assert bot.team == ALLY, "You can only view info from an ally bot perspective"
        try:
            return bot.info_from_my_perspective[self.bot_id]
        except KeyError:
            raise ValueError("The bot on which you want info died after last turn (so no info available)," +
                             " he was alive and targeted last turn which is why it appears in your code")

    def get_closest_enemy(self):
        assert self.team == ALLY, "You can only call closestEnemy function on an ally bot"
        return min(enemyBots, key=lambda bot: bot.viewed_by(self).dist_bot_rank)

    def is_dead(self):
        return self.health == -1


def main():
    # game loop
    while True:
        global orders, enemyBots, ally_bots, bots
        enemyBots = []
        ally_bots = []
        bots = {}
        orders = ""
        ally_bot_alive = int(input())  # the amount of your bot which are still alive
        bot_count = int(input())  # the amount of bots in the arena
        for i in range(bot_count):
            # bot_id: the unique bot id, stay the same for the whole game
            # bot_type: the bot type in a string. It can be ALLY | ENEMY
            # health: the approximate bot health. Can be 0 | 25 | 50 | 75 | 100, 25 meaning that your life is >= 25% and < 50% of your max life
            # shield: the approximate bot shield. Can be 0 | 1 | 25 | 50 | 75 | 100, 1 meaning that your shield is >= 1% and < 25% of your max shield and 0 that you have no more shield left
            # action: action executed by the bot last turn
            # targets: list of the targets id targeted by the bot last turn ("id1;id2;id3...") if the bot is a bot, else -1 (the target for IDLE is the bot itself)
            # enemy_range: /!\ALWAYS 0 FOR THE FIRST LEAGUE/!\ then, the RANGE from to the closest enemy (if the bot is an enemy it returns the range to its closest ally)
            # border_range: /!\ALWAYS 0 FOR THE FIRST LEAGUE/!\ approximate distance between the bot and the closest border
            # border_dist_rank: /!\ALWAYS 0 FOR THE FIRST 2 LEAGUES/!\ bot are sorted in ascending order based on their distance to the closest border
            # dist_en_rank: /!\ALWAYS 0 FOR THE FIRST 2 LEAGUES/!\ bots are sorted by ascending order based on their distance to the closest enemy
            # health_rank: /!\ALWAYS 0 FOR THE FIRST 2 LEAGUES/!\ bots are sorted in ascendant order based on their amount of health, this is the rank of the current bot in the sorted list
            # shield_rank: /!\ALWAYS 0 FOR THE FIRST 2 LEAGUES/!\ bots are sorted in ascendant order based on their amount of shield
            # total_rank: /!\ALWAYS 0 FOR THE FIRST 2 LEAGUES/!\ bots are sorted in ascendant order based on their amount of health + shield
            bot_id, bot_type, health, shield, action, targets, enemy_range, border_range, border_dist_rank, dist_en_rank, health_rank, shield_rank, total_rank = input().split()
            bot_id = int(bot_id)
            health = int(health)
            shield = int(shield)
            enemy_range = int(enemy_range)
            border_range = int(border_range)
            border_dist_rank = int(border_dist_rank)
            dist_en_rank = int(dist_en_rank)
            health_rank = int(health_rank)
            shield_rank = int(shield_rank)
            total_rank = int(total_rank)
            bot = Bot(bot_id, bot_type, shield, health, action, targets, enemy_range, dist_en_rank, health_rank,
                      shield_rank, total_rank, border_range, border_dist_rank)
            bots[bot_id] = bot
            if bot_type == ALLY:
                ally_bots.append(bot)
            else:
                enemyBots.append(bot)
        for i in range(ally_bot_alive):
            on_air_id = int(input().split(" ")[0])  # the id of the bot which is on air
            for j in range(bot_count - 1):
                # bot_id: the unique bot id
                # bot_type: the bot type in a string. It can be ALLY | ENEMY
                # range_from_on_air_bot: approximate distance between the target and the ON AIR bot. Can be 0 to 3 for short, medium, long and out of range
                # dist_me_rank: bots are sorted by ascending order based on their distance to the ON AIR bot
                # shield_comp: /!\ALWAYS 0 FOR THE FIRST LEAGUE/!\ -1 if the bot has more shield than the ON AIR bot, 0 if it's equal, 1 if the ON AIR bot has more shield
                # health_comp: /!\ALWAYS 0 FOR THE FIRST LEAGUE/!\ same as shieldComp but for the health
                # tot_comp: /!\ALWAYS 0 FOR THE FIRST LEAGUE/!\ same as shieldComp but based on the sum of health+shield
                bot_id, _, range_from_on_air_bot, dist_me_rank, shield_comp, health_comp, tot_comp = input().split()
                bot_id = int(bot_id)
                range_from_on_air_bot = int(range_from_on_air_bot)
                dist_me_rank = int(dist_me_rank)
                shield_comp = int(shield_comp)
                health_comp = int(health_comp)
                tot_comp = int(tot_comp)
                info = BotInfo(range_from_on_air_bot, dist_me_rank, shield_comp, health_comp, tot_comp)
                bots[on_air_id].add_info_from_my_perspective(bot_id, info)

        for ally_bot in ally_bots:
            ally_bot.move(ally_bot.get_closest_enemy())

        # Write an answer using print
        # To debug: print("Debug messages...", file=sys.stderr)

        print(orders)


main()
